File: tests/netDxf.Conformance/RawLegacyProfileTests.cs
using System.Security.Cryptography;
using System.Text;
using netDxf;
using netDxf.Header;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static readonly (string File, DxfVersion Version, int CodePage, int Tags, string Sha)[] LegacyFixtures =
    {
        ("small_r13.dxf", DxfVersion.AutoCad13, 932, 1502, "a1007452f8ada2e71fdef60f79923924045cb371f37eeab896a61f5628573d0b"),
        ("small_r14.dxf", DxfVersion.AutoCad14, 1252, 789, "92e792f8f9169e4226c9b7dbe848ba9afb96d3e388e6297c55481e1941c65fb0"),
        ("bin_dxf_r13.dxf", DxfVersion.AutoCad13, 1252, 2075, "c65782683815ff0f1680d26207da1637a486f96b30bc215685e26cf7adee8464"),
        ("bin_dxf_r14.dxf", DxfVersion.AutoCad14, 1252, 2083, "177a97e5dc86ca106d0dd6365d139122fc54ff78df87670847eb7174a8bbd5eb")
    };

    private static void RegisterRawLegacyProfileTests()
    {
        foreach (var fixture in LegacyFixtures)
        {
            var f = fixture;
            Run($"raw-legacy/external/{f.File}", () => RawLegacyExternal(f));
        }
        foreach (DxfVersion version in new[] { DxfVersion.AutoCad13, DxfVersion.AutoCad14 })
            foreach (bool binary in new[] { false, true })
            {
                DxfVersion v = version; bool b = binary;
                Run($"raw-legacy/authored/{v}/{b}", () => RawLegacyAuthored(v, b));
                Run($"raw-legacy/invalid-and-bounded/{v}/{b}", () => RawLegacyBoundaries(v, b));
            }
        foreach (DxfVersion version in new[] { DxfVersion.AutoCad13, DxfVersion.AutoCad14, DxfVersion.AutoCad2000, DxfVersion.AutoCad2004 })
            foreach (bool binary in new[] { false, true })
                foreach (var pair in new[] { ("dos437", 437, "é"), ("DoS850", 850, "é"), ("dos932", 932, "東京") })
                {
                    DxfVersion v = version; bool b = binary; var p = pair;
                    Run($"raw-legacy/dos-codepage/{v}/{b}/{p.Item1}", () => RawLegacyEncoding(v, b, p.Item1, p.Item2, p.Item3));
                }
        Run("raw-legacy/typed-admission-unchanged", RawLegacyTypedBoundary);
        Run("raw-legacy/invalid-codepage-aliases", RawLegacyBadAliases);
    }

    private static List<DxfTag> RawLegacyTags(DxfVersion version, string codePage = "ANSI_1252", string text = "legacy text")
        => new()
        {
            new(0, "SECTION"), new(2, "HEADER"), new(9, "$ACADVER"),
            new(1, version switch { DxfVersion.AutoCad13 => "AC1012", DxfVersion.AutoCad14 => "AC1014", _ => HeaderVersion(version) }),
            new(9, "$DWGCODEPAGE"), new(3, codePage), new(0, "ENDSEC"),
            new(0, "SECTION"), new(2, "ENTITIES"), new(0, "LINE"), new(5, "A"),
            new(100, "AcDbEntity"), new(8, "0"), new(100, "AcDbLine"),
            new(10, 1e-20), new(20, -0.0), new(30, 3.0), new(11, 4.0), new(21, 5.0), new(31, 6.0),
            new(1001, "RAW_TEST"), new(1000, text), new(1004, new byte[] { 0, 128, 255 }),
            new(1005, "B"), new(1070, (short)-1), new(1071, int.MinValue), new(0, "ENDSEC"), new(0, "EOF")
        };

    private static void RawLegacyExternal((string File, DxfVersion Version, int CodePage, int Tags, string Sha) fixture)
    {
        byte[] bytes = File.ReadAllBytes(Path.Combine("tests", "fixtures", "legacy", fixture.File));
        Equal(fixture.Sha, Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant(), "Pinned external fixture hash");
        var raw = LoadRaw(bytes);
        Equal(fixture.Version, raw.Version, "External declared version");
        Equal(fixture.CodePage, raw.EncodingCodePage, "External encoding");
        Equal(fixture.Tags, raw.Tags.Count, "External ordered tag count");
        Equal(fixture.File.StartsWith("bin_", StringComparison.Ordinal), raw.IsBinary, "External transport");
        Check(bytes.SequenceEqual(SaveRaw(raw)), "External original bytes changed");
        SameRawTags(raw.Tags, LoadRaw(SaveRaw(raw.WithTags(raw.Tags))).Tags);
        SameRawTags(raw.Tags, LoadRaw(SaveRaw(raw, !raw.IsBinary)).Tags);
        CheckRawRecordPartition(raw);
        var layer = raw.Sections.Single(s => s.Name == "TABLES").Records.First(r => r.Name == "LAYER");
        Check(layer.Content.Any(t => t.Code == 62), "Expected an authored layer color");
        var replacement = layer.Tags.Select(t => t.Code == 62 ? new DxfTag(62, (short)3) : t).ToArray();
        var edited = raw.WithRecord(layer, replacement);
        AssertOutsideRecordUnchanged(raw, layer, edited, replacement.Length);
        SameRawTags(edited.Tags, LoadRaw(SaveRaw(edited, !raw.IsBinary)).Tags);
        // Independent tooling gets normalized text with the declared encoding unchanged.
        File.WriteAllBytes(Path.Combine(ArtifactDirectory, "legacy-normalized-" + fixture.File), SaveRaw(raw.WithTags(raw.Tags), false));
        File.WriteAllBytes(Path.Combine(ArtifactDirectory, "legacy-edited-" + fixture.File), SaveRaw(edited, false));
    }

    private static void RawLegacyAuthored(DxfVersion version, bool binary)
    {
        var tags = RawLegacyTags(version);
        var raw = LoadRaw(RawFixtureBytes(tags, binary));
        Equal(version, raw.Version, "Authored legacy profile");
        SameRawTags(tags, raw.Tags);
        SameRawTags(tags, LoadRaw(SaveRaw(DxfRawDocument.Create(tags, binary), !binary)).Tags);
        SameRawTags(tags, LoadRaw(SaveRaw(raw.WithTags(raw.Tags))).Tags);
        CheckRawRecordPartition(raw);
    }

    private static void RawLegacyBoundaries(DxfVersion version, bool binary)
    {
        var tags = RawLegacyTags(version); byte[] bytes = RawFixtureBytes(tags, binary);
        Throws<InvalidDataException>(() => LoadRaw(bytes, new DxfRawOptions(maximumBytes: bytes.Length - 1)));
        Throws<InvalidDataException>(() => LoadRaw(bytes, new DxfRawOptions(maximumTags: tags.Count - 1)));
        Throws<InvalidDataException>(() => LoadRaw(bytes, new DxfRawOptions(maximumStringLength: 9)));
        var raw = LoadRaw(bytes);
        Throws<NotSupportedException>(() => raw.WithTags(tags.Select(t => t.Code == 1 ? new DxfTag(1, "AC1015") : t)));
        using var cancelled = new MemoryStream(bytes);
        Throws<OperationCanceledException>(() => DxfRawDocument.Load(cancelled, cancellationToken: new CancellationToken(true)));
        Check(cancelled.CanRead, "Legacy cancellation closed input");
        foreach (string unsupported in new[] { "AC1009", "AC1034", "VENDOR_UNKNOWN" })
        {
            var changed = tags.Select(t => t.Code == 1 ? new DxfTag(1, unsupported) : t);
            Throws<DxfVersionNotSupportedException>(() => LoadRaw(RawFixtureBytes(changed, binary)));
        }
        byte[] truncated = RawFixtureBytes(tags.Take(tags.Count - 1), binary);
        Throws<EndOfStreamException>(() => LoadRaw(truncated));
    }

    private static void RawLegacyEncoding(DxfVersion version, bool binary, string name, int codePage, string text)
    {
        var tags = RawLegacyTags(version, name, text);
        var encoding = Encoding.GetEncoding(codePage, EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback);
        var raw = LoadRaw(RawFixtureBytes(tags, binary, encoding));
        Equal(codePage, raw.EncodingCodePage, "DOS alias resolution");
        SameRawTags(tags, raw.Tags);
        SameRawTags(tags, LoadRaw(SaveRaw(raw.WithTags(raw.Tags), !binary)).Tags);
        Equal(name, (string)raw.Tags.Single(t => t.Code == 3).Value, "Alias spelling changed");
    }

    private static void RawLegacyTypedBoundary()
    {
        foreach (DxfVersion version in new[] { DxfVersion.AutoCad13, DxfVersion.AutoCad14 })
        {
            Throws<NotSupportedException>(() => new DxfDocument(version));
            foreach (bool binary in new[] { false, true })
            {
                using var input = new MemoryStream(RawFixtureBytes(RawLegacyTags(version), binary));
                Throws<DxfVersionNotSupportedException>(() => DxfDocument.Load(input));
                Check(input.CanRead, "Unsupported typed load closed input");
            }
        }
    }

    private static void RawLegacyBadAliases()
    {
        foreach (string name in new[] { "dos", "dos_932", "dos+932", "dos 932", "dos932 ", "dos99999999999", "UTF-16", "ANSI_" })
            Throws<NotSupportedException>(() => DxfRawDocument.Create(RawLegacyTags(DxfVersion.AutoCad13, name)));
        foreach (string name in new[] { "dos1200", "dos1201", "dos12000", "dos12001" })
            Throws<NotSupportedException>(() => DxfRawDocument.Create(RawLegacyTags(DxfVersion.AutoCad13, name)));
    }
}

File: javascript/tests/netDxf.Conformance/RawLegacyProfileTests.js
// Port of RawLegacyProfileTests.cs. RawLegacyTypedBoundary awaits the typed JavaScript engine.
import fs from 'node:fs';
import path from 'node:path';
import { createHash } from 'node:crypto';
import { sourceRoot, javascriptRoot } from '../../tools/dotnet.mjs';
import { DxfTag, DxfRawDocument, DxfRawOptions, DxfVersion, DxfVersionNotSupportedException, MemoryStream } from '../../index.js';
import { Encoding } from '../../runtime/Encoding.js';
import * as E from '../../runtime/Errors.js';
import { Run, Check, Equal, Throws, VersionName, HeaderVersion, BooleanName } from './TestHarness.js';
import { RawFixtureBytes, LoadRaw, SaveRaw, SameRawTags } from './RawDocumentTests.js';
import { CheckRawRecordPartition, AssertOutsideRecordUnchanged } from './RawRecordTests.js';
const T=(code,value)=>new DxfTag(code,value);
export const LegacyFixtures=Object.freeze([
  ['small_r13.dxf',DxfVersion.AutoCad13,932,1502,'a1007452f8ada2e71fdef60f79923924045cb371f37eeab896a61f5628573d0b'],
  ['small_r14.dxf',DxfVersion.AutoCad14,1252,789,'92e792f8f9169e4226c9b7dbe848ba9afb96d3e388e6297c55481e1941c65fb0'],
  ['bin_dxf_r13.dxf',DxfVersion.AutoCad13,1252,2075,'c65782683815ff0f1680d26207da1637a486f96b30bc215685e26cf7adee8464'],
  ['bin_dxf_r14.dxf',DxfVersion.AutoCad14,1252,2083,'177a97e5dc86ca106d0dd6365d139122fc54ff78df87670847eb7174a8bbd5eb'],
].map(Object.freeze));
const WriteArtifact=(name,bytes)=>{const dir=process.env.DXF_JS_TEST_ARTIFACTS||path.join(javascriptRoot,'artifacts/conformance');fs.mkdirSync(dir,{recursive:true});fs.writeFileSync(path.join(dir,name),bytes);};
export function RegisterRawLegacyProfileTests(){
  for(const f of LegacyFixtures)Run(`raw-legacy/external/${f[0]}`,()=>RawLegacyExternal(f));
  for(const v of [DxfVersion.AutoCad13,DxfVersion.AutoCad14])for(const b of [false,true]){
    Run(`raw-legacy/authored/${VersionName(v)}/${BooleanName(b)}`,()=>RawLegacyAuthored(v,b));
    Run(`raw-legacy/invalid-and-bounded/${VersionName(v)}/${BooleanName(b)}`,()=>RawLegacyBoundaries(v,b));
  }
  for(const v of [DxfVersion.AutoCad13,DxfVersion.AutoCad14,DxfVersion.AutoCad2000,DxfVersion.AutoCad2004])
    for(const b of [false,true])for(const p of [['dos437',437,'é'],['DoS850',850,'é'],['dos932',932,'東京']])
      Run(`raw-legacy/dos-codepage/${VersionName(v)}/${BooleanName(b)}/${p[0]}`,()=>RawLegacyEncoding(v,b,...p));
  Run('raw-legacy/invalid-codepage-aliases',RawLegacyBadAliases);
}
export function RawLegacyTags(version,codePage='ANSI_1252',text='legacy text'){
  return [T(0,'SECTION'),T(2,'HEADER'),T(9,'$ACADVER'),T(1,HeaderVersion(version)),T(9,'$DWGCODEPAGE'),T(3,codePage),T(0,'ENDSEC'),
    T(0,'SECTION'),T(2,'ENTITIES'),T(0,'LINE'),T(5,'A'),T(100,'AcDbEntity'),T(8,'0'),T(100,'AcDbLine'),
    T(10,1e-20),T(20,-0),T(30,3),T(11,4),T(21,5),T(31,6),T(1001,'RAW_TEST'),T(1000,text),T(1004,Uint8Array.of(0,128,255)),
    T(1005,'B'),T(1070,-1),T(1071,-2147483648),T(0,'ENDSEC'),T(0,'EOF')];
}
export function RawLegacyExternal([file,version,codePage,count,hash]){
  const bytes=new Uint8Array(fs.readFileSync(path.join(sourceRoot,'tests/fixtures/legacy',file)));
  Equal(hash,createHash('sha256').update(bytes).digest('hex'));
  const raw=LoadRaw(bytes);Equal(version,raw.Version);Equal(codePage,raw.EncodingCodePage);Equal(count,raw.Tags.Count);
  Equal(file.startsWith('bin_'),raw.IsBinary);Equal(bytes,SaveRaw(raw));
  SameRawTags(raw.Tags,LoadRaw(SaveRaw(raw.WithTags(raw.Tags))).Tags);SameRawTags(raw.Tags,LoadRaw(SaveRaw(raw,!raw.IsBinary)).Tags);
  CheckRawRecordPartition(raw);
  const layer=raw.Sections.find(s=>s.Name==='TABLES').Records.find(r=>r.Name==='LAYER');Check([...layer.Content].some(t=>t.Code===62));
  const replacement=[...layer.Tags].map(t=>t.Code===62?T(62,3):t),edited=raw.WithRecord(layer,replacement);
  AssertOutsideRecordUnchanged(raw,layer,edited,replacement.length);SameRawTags(edited.Tags,LoadRaw(SaveRaw(edited,!raw.IsBinary)).Tags);
  WriteArtifact('legacy-normalized-'+file,SaveRaw(raw.WithTags(raw.Tags),false));WriteArtifact('legacy-edited-'+file,SaveRaw(edited,false));
}
export function RawLegacyAuthored(version,binary){
  const tags=RawLegacyTags(version),raw=LoadRaw(RawFixtureBytes(tags,binary));Equal(version,raw.Version);SameRawTags(tags,raw.Tags);
  SameRawTags(tags,LoadRaw(SaveRaw(DxfRawDocument.Create(tags,binary),!binary)).Tags);
  SameRawTags(tags,LoadRaw(SaveRaw(raw.WithTags(raw.Tags))).Tags);CheckRawRecordPartition(raw);
}
export function RawLegacyBoundaries(version,binary){
  const tags=RawLegacyTags(version),bytes=RawFixtureBytes(tags,binary);
  Throws(E.InvalidDataException,()=>LoadRaw(bytes,new DxfRawOptions(bytes.length-1)));
  Throws(E.InvalidDataException,()=>LoadRaw(bytes,new DxfRawOptions(undefined,tags.length-1)));
  Throws(E.InvalidDataException,()=>LoadRaw(bytes,new DxfRawOptions(undefined,undefined,9)));
  const raw=LoadRaw(bytes);Throws(E.NotSupportedException,()=>raw.WithTags(tags.map(t=>t.Code===1?T(1,'AC1015'):t)));
  const cancelled=new MemoryStream(bytes);Throws(E.OperationCanceledException,()=>DxfRawDocument.Load(cancelled,null,{aborted:true}));Check(cancelled.CanRead);
  for(const version of ['AC1006','AC1034','VENDOR_UNKNOWN'])
    Throws(DxfVersionNotSupportedException,()=>LoadRaw(RawFixtureBytes(tags.map(t=>t.Code===1?T(1,version):t),binary)));
  Throws(E.EndOfStreamException,()=>LoadRaw(RawFixtureBytes(tags.slice(0,-1),binary)));
}
export function RawLegacyEncoding(version,binary,name,codePage,text){
  const tags=RawLegacyTags(version,name,text),raw=LoadRaw(RawFixtureBytes(tags,binary,Encoding.GetEncoding(codePage)));
  Equal(codePage,raw.EncodingCodePage);SameRawTags(tags,raw.Tags);SameRawTags(tags,LoadRaw(SaveRaw(raw.WithTags(raw.Tags),!binary)).Tags);
  Equal(name,raw.Tags.find(t=>t.Code===3).Value);
}
export function RawLegacyBadAliases(){
  for(const name of ['dos','dos_932','dos+932','dos 932','dos932 ','dos99999999999','UTF-16','ANSI_',
    'dos1200','dos1201','dos12000','dos12001'])
    Throws(E.NotSupportedException,()=>DxfRawDocument.Create(RawLegacyTags(DxfVersion.AutoCad13,name)));
}
